tools-local/tasks/mobile.tasks/WasmAppBuilder/IcallTableGenerator.cs

```diff
@@ -0,0 +1,301 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Text.Json;
+using System.Reflection;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+public class IcallTableGenerator : Task
+{
+    [Required]
+    public string? RuntimeIcallTableFile { get; set; }
+    [Required]
+    public ITaskItem[]? Assemblies { get; set; }
+    [Required]
+    public string? OutputPath { get; set; }
+
+    private List<Icall> _icalls = new List<Icall> ();
+    private Dictionary<string, IcallClass> _runtimeIcalls = new Dictionary<string, IcallClass> ();
+
+    public override bool Execute()
+    {
+        Log.LogMessage(MessageImportance.Normal, $"Generating icall table to '{OutputPath}'.");
+        GenIcallTable(RuntimeIcallTableFile!, Assemblies!.Select(item => item.ItemSpec).ToArray());
+        return true;
+    }
+
+    //
+    // Given the runtime generated icall table, and a set of assemblies, generate
+    // a smaller linked icall table mapping tokens to C function names
+    // The runtime icall table should be generated using
+    // mono --print-icall-table
+    //
+    public void GenIcallTable(string runtimeIcallTableFile, string[] assemblies)
+    {
+        ReadTable (runtimeIcallTableFile);
+        var resolver = new PathAssemblyResolver(assemblies);
+        var mlc = new MetadataLoadContext(resolver, "System.Private.CoreLib");
+        foreach (var aname in assemblies)
+        {
+            var a = mlc.LoadFromAssemblyPath(aname);
+            foreach (var type in a.GetTypes())
+                ProcessType(type);
+        }
+
+        using (var w = File.CreateText(OutputPath!))
+            EmitTable (w);
+    }
+
+    private void EmitTable (StreamWriter w)
+    {
+        var assemblyMap = new Dictionary<string, string> ();
+        foreach (var icall in _icalls)
+            assemblyMap [icall.Assembly!] = icall.Assembly!;
+
+        foreach (var assembly in assemblyMap.Keys)
+        {
+            var sorted = _icalls.Where (i => i.Assembly == assembly).ToArray ();
+            Array.Sort (sorted);
+
+            string aname;
+            if (assembly == "mscorlib" || assembly == "System.Private.CoreLib")
```

Comment:
`mscorlib` is a pure facade in dotnet/runtime, it will never have icalls.

Reply:
Its a leftover from the old code, will fix.